overlord/devicestate/devicestate.go

```diff
@@ -967,39 +941,31 @@ func remodelTasks(ctx context.Context, st *state.State, current, new *asserts.Mo
 			// revision was unchanged. in this case, we treat the snap as if it
 			// were unchanged.
 			if ts.MaybeEdge(snapstate.LastBeforeLocalModificationsEdge) != nil {
-				if err := updateNeededSnapsFromTs(ts); err != nil {
-					return nil, err
-				}
-
 				continue
 			}
 		}
 
 		// if we're here, the snap that is installed is unchanged from the snap
 		// that the model requires. the snap may have had a channel change, but
 		// that channel change did not result in a revision change.
+		tracker.Add(currentInfo)
+	}
 
-		if currentInfo.Base != "" {
-			neededSnaps[currentInfo.Base] = true
-		}
-		// deal with content providers
-		for defProvider := range snap.NeededDefaultProviders(currentInfo) {
-			neededSnaps[defProvider] = true
-		}
+	warnings, errs := tracker.Check()
+	for _, w := range warnings {
+		logger.Noticef("remodel prerequisites warning: %v", w)
 	}
-	// Now we know what snaps are in the model and whether they have any
-	// dependencies. Verify that the model is self contained, in the sense
-	// that all prerequisites of the snaps in the model, i.e. bases and
-	// default content providers are explicitly listed in the model
-	var missingSnaps []string
-	for needed := range neededSnaps {
-		if !snapsAccountedFor[needed] {
-			missingSnaps = append(missingSnaps, needed)
+
+	if len(errs) > 0 {
+		var builder strings.Builder
+		builder.WriteString("cannot remodel to model that is not self contained:")
+
+		for _, err := range errs {
+			builder.WriteString("\n  - ")
+			builder.WriteString(err.Error())
 		}
-	}
-	if len(missingSnaps) != 0 {
-		sort.Strings(missingSnaps)
-		return nil, fmt.Errorf("cannot remodel with incomplete model, the following snaps are required but not listed: %s", strutil.Quoted(missingSnaps))
+
+		return nil, errors.New(builder.String())
 	}
 
 	// TODO: fix the prerequisite task getting stuck during a remodel by
```

Comment:
I think this PR partially addresses this as the the task created won't have prereqs anymore except for bases? what's the status with respect of installing bases before dependent snaps in a remodel?

Reply:
Bases will be installed first due to this change that was made a little while back: https://github.com/snapcore/snapd/blob/master/overlord/devicestate/devicestate.go?plain=1#L872

Reply:
We just install all bases first, regardless of dependencies.

Reply:
you can't run hooks without the base snap being already installed, it seems we might have done enough changes now that this TODO is covered and any further issues would be a bug?